
# Data Cleaning and KNN Algorithm Implementation without Sklearn


 # Task 1 -Data Preparation

In [1]:
import pandas as pd
import numpy as np
from sklearn import preprocessing
import warnings
warnings.filterwarnings("ignore") # Turning off the warnings
dataq1 = pd.read_csv('Q1',names=['A1','A2','A3','A4','A5','A6','A7','A8','A9','A10','A11','A12','A13','A14','A15','A16'],index_col=False)
dataq1

A1     A2      A3 A4 A5  A6  A7      A8 A9 A10  A11 A12 A13    A14    A15  \
0    b  30.83   0.000  u  g   w   v   1.250  t   t    1   f   g  00202      0   
1    a  58.67   4.460  u  g   q   h   3.040  t   t    6   f   g  00043    560   
2    a  24.50   0.500  u  g   q   h   1.500  t   f    0   f   g  00280    824   
3    b  27.83   1.540  u  g   w   v   3.750  t   t    5   t   g  00100      3   
4    b  20.17   5.625  u  g   w   v   1.710  t   f    0   f   s  00120      0   
5    b  32.08   4.000  u  g   m   v   2.500  t   f    0   t   g  00360      0   
6    b  33.17   1.040  u  g   r   h   6.500  t   f    0   t   g  00164  31285   
7    a  22.92  11.585  u  g  cc   v   0.040  t   f    0   f   g  00080   1349   
8    b  54.42   0.500  y  p   k   h   3.960  t   f    0   f   g  00180    314   
9    b  42.50   4.915  y  p   w   v   3.165  t   f    0   t   g  00052   1442   
10   b  22.08   0.830  u  g   c   h   2.165  f   f    0   t   g  00128      0   
11   b  29.92   1.835  u  g   c   h   4.335  t   f    0   f   g  00260    200   
12   a  38.25   6.000  u  g   k   v   1.000  t   f    0   t   g  00000      0   
13   b  48.08   6.040  u  g   k   v   0.040  f   f    0   f   g  00000   2690   
14   a  45.83  10.500  u  g   q   v   5.000  t   t    7   t   g  00000      0   
15   b  36.67   4.415  y  p   k   v   0.250  t   t   10   t   g  00320      0   
16   b  28.25   0.875  u  g   m   v   0.960  t   t    3   t   g  00396      0   
17   a  23.25   5.875  u  g   q   v   3.170  t   t   10   f   g  00120    245   
18   b  21.83   0.250  u  g   d   h   0.665  t   f    0   t   g  00000      0   
19   a  19.17   8.585  u  g  cc   h   0.750  t   t    7   f   g  00096      0   
20   b  25.00  11.250  u  g   c   v   2.500  t   t   17   f   g  00200   1208   
21   b  23.25   1.000  u  g   c   v   0.835  t   f    0   f   s  00300      0   
22   a  47.75   8.000  u  g   c   v   7.875  t   t    6   t   g  00000   1260   
23   a  27.42  14.500  u  g   x   h   3.085  t   t    1   f   g  00120     11   
24   a  41.17   6.500  u  g   q   v   0.500  t   t    3   t   g  00145      0   
25   a  15.83   0.585  u  g   c   h   1.500  t   t    2   f   g  00100      0   
26   a  47.00  13.000  u  g   i  bb   5.165  t   t    9   t   g  00000      0   
27   b  56.58  18.500  u  g   d  bb  15.000  t   t   17   t   g  00000      0   
28   b  57.42   8.500  u  g   e   h   7.000  t   t    3   f   g  00000      0   
29   b  42.08   1.040  u  g   w   v   5.000  t   t    6   t   g  00500  10000   
..  ..    ...     ... .. ..  ..  ..     ... ..  ..  ...  ..  ..    ...    ...   
660  b  22.25   9.000  u  g  aa   v   0.085  f   f    0   f   g  00000      0   
661  b  29.83   3.500  u  g   c   v   0.165  f   f    0   f   g  00216      0   
662  a  23.50   1.500  u  g   w   v   0.875  f   f    0   t   g  00160      0   
663  b  32.08   4.000  y  p  cc   v   1.500  f   f    0   t   g  00120      0   
664  b  31.08   1.500  y  p   w   v   0.040  f   f    0   f   s  00160      0   
665  b  31.83   0.040  y  p   m   v   0.040  f   f    0   f   g  00000      0   
666  a  21.75  11.750  u  g   c   v   0.250  f   f    0   t   g  00180      0   
667  a  17.92   0.540  u  g   c   v   1.750  f   t    1   t   g  00080      5   
668  b  30.33   0.500  u  g   d   h   0.085  f   f    0   t   s  00252      0   
669  b  51.83   2.040  y  p  ff  ff   1.500  f   f    0   f   g  00120      1   
670  b  47.17   5.835  u  g   w   v   5.500  f   f    0   f   g  00465    150   
671  b  25.83  12.835  u  g  cc   v   0.500  f   f    0   f   g  00000      2   
672  a  50.25   0.835  u  g  aa   v   0.500  f   f    0   t   g  00240    117   
673  ?  29.50   2.000  y  p   e   h   2.000  f   f    0   f   g  00256     17   
674  a  37.33   2.500  u  g   i   h   0.210  f   f    0   f   g  00260    246   
675  a  41.58   1.040  u  g  aa   v   0.665  f   f    0   f   g  00240    237   
676  a  30.58  10.665  u  g   q   h   0.085  f   t   12   t   g  00129      3   
677  b  19.42   7.250  u  g   m   v

In [2]:
dataq1.isna().sum() # Checking missing values

A1     0
A2     0
A3     0
A4     0
A5     0
A6     0
A7     0
A8     0
A9     0
A10    0
A11    0
A12    0
A13    0
A14    0
A15    0
A16    0
dtype: int64

### Although it shows that number of missing values are 0 but the data description file shows that there are missing values in the dataset.

In [3]:
print(dataq1['A1'].value_counts())
print(dataq1['A2'].value_counts())
print(dataq1['A4'].value_counts())
print(dataq1['A5'].value_counts())
print(dataq1['A6'].value_counts())
print(dataq1['A7'].value_counts())

b    468
a    210
?     12
Name: A1, dtype: int64
?        12
22.67     9
20.42     7
23.58     6
22.50     6
25.00     6
20.67     6
19.17     6
24.50     6
18.83     6
23.00     5
33.17     5
27.67     5
23.08     5
27.83     5
23.25     5
22.08     4
27.25     4
20.00     4
24.75     4
41.17     4
32.33     4
26.17     4
28.58     4
26.67     4
22.92     4
25.67     4
35.17     4
25.17     4
29.50     4
         ..
44.00     1
36.00     1
44.83     1
30.33     1
56.75     1
17.33     1
35.42     1
24.92     1
17.42     1
26.58     1
37.42     1
27.33     1
56.50     1
16.17     1
55.92     1
21.42     1
62.50     1
46.08     1
26.00     1
54.33     1
28.08     1
38.42     1
52.17     1
54.42     1
47.25     1
74.83     1
45.33     1
58.67     1
54.83     1
15.75     1
Name: A2, Length: 350, dtype: int64
u    519
y    163
?      6
l      2
Name: A4, dtype: int64
g     519
p     163
?       6
gg      2
Name: A5, dtype: int64
c     137
q      78
w      64
i      59
aa     54
ff     53


### The above output shows that there are 12, 12, 6, 6, 9, 9, and 13 '?'s in A1, A2, A4, A5, A6, A7 and A14. These counts are the same as counts for missing values in data description file. 

In [4]:
# Changing '?' to NaN and checking number of Na values
ndata = dataq1 # Creating copy of dataset
ndata[['A1','A2','A4','A5','A6','A7','A14']] = ndata[['A1','A2','A4','A5','A6','A7','A14']].replace('?', np.NaN) # Replacing ? with NA
ndata.isna().sum() # Counting number of NA values

A1     12
A2     12
A3      0
A4      6
A5      6
A6      9
A7      9
A8      0
A9      0
A10     0
A11     0
A12     0
A13     0
A14    13
A15     0
A16     0
dtype: int64

### Checking datatypes of attributes

In [5]:
ndata.dtypes

A1      object
A2      object
A3     float64
A4      object
A5      object
A6      object
A7      object
A8     float64
A9      object
A10     object
A11      int64
A12     object
A13     object
A14     object
A15      int64
A16     object
dtype: object

### A2, A3, A8, A11, A14, and A15 should be continuous values. Therefore, typecasting A2, and A14

In [6]:
ndata['A2'] = pd.to_numeric(ndata['A2']);
ndata['A14'] = pd.to_numeric(ndata['A14']);
print(ndata['A2'].dtype)
print(ndata['A14'].dtype)

float64
float64


### Among missing value attributes only A2 and A14 are continuous numeric variable. Imputing them with median values.

In [7]:
ndata['A2'].fillna(ndata['A2'].median(), inplace = True)
ndata['A14'].fillna(ndata['A14'].median(), inplace = True)
ndata.isna().sum() # Verifying if missing values are removed across A2 and A14

A1     12
A2      0
A3      0
A4      6
A5      6
A6      9
A7      9
A8      0
A9      0
A10     0
A11     0
A12     0
A13     0
A14     0
A15     0
A16     0
dtype: int64

### Imputing Categorical variable using Mode

In [8]:
# Imputing by mode of categorical variables
ndata['A1'].fillna(ndata['A1'].mode()[0], inplace=True)
ndata['A4'].fillna(ndata['A4'].mode()[0], inplace=True)
ndata['A5'].fillna(ndata['A5'].mode()[0], inplace=True)
ndata['A6'].fillna(ndata['A6'].mode()[0], inplace=True)
ndata['A7'].fillna(ndata['A7'].mode()[0], inplace=True)

### The above output shows that all missing values are now handled.

### Discretizing and integer encoding numerical descriptive feature 'A2' 

In [9]:
ndata1 = ndata.copy() # Creating copy of a dataset
ndata1['A2'] = pd.qcut(ndata1['A2'], q=3, labels=['low', 'medium', 'high']) # Performing discretization                                    

In [10]:
ndata1['A2'].value_counts() # Checking counts of dicretized levels

medium    231
low       230
high      229
Name: A2, dtype: int64

In [11]:
ndata1.head(5)

A1      A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b  medium  0.000  u  g  w  v  1.25  t   t    1   f   g  202.0    0   +
1  a    high  4.460  u  g  q  h  3.04  t   t    6   f   g   43.0  560   +
2  a  medium  0.500  u  g  q  h  1.50  t   f    0   f   g  280.0  824   +
3  b  medium  1.540  u  g  w  v  3.75  t   t    5   t   g  100.0    3   +
4  b     low  5.625  u  g  w  v  1.71  t   f    0   f   s  120.0    0   +

### 'A2' feature is successfully discretized

### Integer Coding 'A2'

In [12]:
level_mapping = {'low': 0, 'medium': 1, 'high': 2} # Defining mapping between levels and integer

In [13]:
ndata1_integer = ndata1.copy() # Creating copy of dataset
ndata1_integer['A2'] = ndata1_integer['A2'].replace(level_mapping) # Replacing levels with integer encoding
ndata1_integer.head(5)

A1  A2     A3 A4 A5 A6 A7    A8 A9 A10  A11 A12 A13    A14  A15 A16
0  b   1  0.000  u  g  w  v  1.25  t   t    1   f   g  202.0    0   +
1  a   2  4.460  u  g  q  h  3.04  t   t    6   f   g   43.0  560   +
2  a   1  0.500  u  g  q  h  1.50  t   f    0   f   g  280.0  824   +
3  b   1  1.540  u  g  w  v  3.75  t   t    5   t   g  100.0    3   +
4  b   0  5.625  u  g  w  v  1.71  t   f    0   f   s  120.0    0   +

### 'A2' successfully Integer encoded

### Encoding objects type variable of the descriptive data to feed them to the SK learn module

In [14]:
ndata1_integer_data = ndata1_integer.drop(columns = 'A16') # Creating only descriptive dataset
ndata1_integer_data
target = ndata1_integer['A16'] # Defining target feature

In [15]:
categorical_cols = ndata1_integer_data.columns[ndata1_integer_data.dtypes==object].tolist() # Creating list of object type variables

In [16]:
categorical_cols # object variables for on-hot-encoding

['A1', 'A4', 'A5', 'A6', 'A7', 'A9', 'A10', 'A12', 'A13']

In [17]:
# Defining binary variable for categorical descriptive feature with only 2 levels,
for col in categorical_cols:
    n = len(ndata1_integer_data[col].unique())
    if (n == 2):
        ndata1_integer_data[col] = pd.get_dummies(ndata1_integer_data[col], drop_first=True)

# using regular one-hot-encoding categorical features (with > 2 levels), making sure that numeric feature is untouched
ndata1_integer_data = pd.get_dummies(ndata1_integer_data)
ndata1_integer_data.head(5)

A1  A2     A3    A8  A9  A10  A11  A12    A14  A15  ...    A7_ff  A7_h  \
0   1   1  0.000  1.25   1    1    1    0  202.0    0  ...        0     0   
1   0   2  4.460  3.04   1    1    6    0   43.0  560  ...        0     1   
2   0   1  0.500  1.50   1    0    0    0  280.0  824  ...        0     1   
3   1   1  1.540  3.75   1    1    5    1  100.0    3  ...        0     0   
4   1   0  5.625  1.71   1    0    0    0  120.0    0  ...        0     0   

   A7_j  A7_n  A7_o  A7_v  A7_z  A13_g  A13_p  A13_s  
0     0     0     0     1     0      1      0      0  
1     0     0     0     0     0      1      0      0  
2     0     0     0     0     0      1      0      0  
3     0     0     0     1     0      1      0      0  
4     0     0     0     1     0      0      0      1  

[5 rows x 42 columns]

### Categorical variables are encoded successfully

In [18]:
# Integer Coding Target Variable
np.unique(target, return_counts=True) # Counting number of instance of each categorical levels

(array(['+', '-'], dtype=object), array([307, 383]))

In [19]:
label_encoder = preprocessing.LabelEncoder() # Using LabelEncoder from Sklearn module to automatically code target feature 
                                             # since we do not have any encoding criteria
le_fit = label_encoder.fit(target)
target_encoded_le = le_fit.transform(target)

In [20]:
print("Target Type:", type(target))
print("Counts Using NumPy:") 
print(np.unique(target_encoded_le, return_counts = True)) # Using Numpy moddule to print value counts of target levels

Target Type: <class 'pandas.core.series.Series'>
Counts Using NumPy:
(array([0, 1]), array([307, 383]))


In [21]:
ndata1_integer_data_scaled = preprocessing.StandardScaler().fit_transform(ndata1_integer_data)
df_clean = pd.DataFrame(ndata1_integer_data_scaled, columns = ndata1_integer_data.columns).round(3)

In [22]:
df_clean['Target'] = target_encoded_le # Naming target feature encoded variable to 'target' and assigning it to final clean dataset

In [23]:
# converting to comma-seperated dataset 
df_clean.to_csv('df_clean.csv', index = False)

In [24]:
df_clean.shape

(690, 43)

In [25]:
df_clean.describe(include='all').round(3)

A1       A2       A3       A8       A9      A10      A11      A12  \
count  690.000  690.000  690.000  690.000  690.000  690.000  690.000  690.000   
mean    -0.000    0.000    0.000    0.000   -0.000    0.000   -0.000    0.000   
std      1.001    1.001    1.001    1.001    1.001    1.001    1.001    1.001   
min     -1.512   -1.224   -0.957   -0.665   -1.048   -0.864   -0.494   -0.919   
25%     -1.512   -1.224   -0.756   -0.616   -1.048   -0.864   -0.494   -0.919   
50%      0.661    0.002   -0.404   -0.366    0.955   -0.864   -0.494   -0.919   
75%      0.661    1.228    0.493    0.120    0.955    1.157    0.123    1.088   
max      0.661    1.228    4.672    7.858    0.955    1.157   13.294    1.088   

           A14      A15   ...        A7_h     A7_j     A7_n     A7_o     A7_v  \
count  690.000  690.000   ...     690.000  690.000  690.000  690.000  690.000   
mean    -0.000    0.000   ...       0.000    0.000    0.000   -0.000   -0.000   
std      1.001    1.001   ...       1.001    1.001    1.001    1.001    1.001   
min     -1.067   -0.195   ...      -0.500   -0.108   -0.076   -0.054   -1.203   
25%     -0.602   -0.195   ...      -0.500   -0.108   -0.076   -0.054   -1.203   
50%     -0.137   -0.194   ...      -0.500   -0.108   -0.076   -0.054    0.831   
75%      0.514   -0.119   ...      -0.500   -0.108   -0.076   -0.054    0.831   
max     10.557   19.012   ...       2.000    9.233   13.096   18.547    0.831   

          A7_z    A13_g    A13_p    A13_s   Target  
count  690.000  690.000  690.000  690.000  690.000  
mean     0.000   -0.000    0.000    0.000    0.555  
std      1.001    1.001    1.001    1.001    0.497  
min     -0.108   -3.101   -0.108   -0.300    0.000  
25%     -0.108    0.322   -0.108   -0.300    0.000  
50%     -0.108    0.322   -0.108   -0.300    1.000  
75%     -0.108    0.322   -0.108   -0.300    1.000  
max      9.233    0.322    9.233    3.332    1.000  

[8 rows x 43 columns]

In [26]:
df_clean.head(5)

A1     A2     A3     A8     A9    A10    A11    A12    A14    A15  \
0  0.661  0.002 -0.957 -0.291  0.955  1.157 -0.288 -0.919  0.107 -0.195   
1 -1.512  1.228 -0.060  0.244  0.955  1.157  0.741 -0.919 -0.817 -0.088   
2 -1.512  0.002 -0.856 -0.216  0.955 -0.864 -0.494 -0.919  0.560 -0.037   
3  0.661  0.002 -0.647  0.457  0.955  1.157  0.535  1.088 -0.486 -0.195   
4  0.661 -1.224  0.174 -0.154  0.955 -0.864 -0.494 -0.919 -0.369 -0.195   

    ...    A7_h   A7_j   A7_n   A7_o   A7_v   A7_z  A13_g  A13_p  A13_s  \
0   ...    -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300   
1   ...     2.0 -0.108 -0.076 -0.054 -1.203 -0.108  0.322 -0.108 -0.300   
2   ...     2.0 -0.108 -0.076 -0.054 -1.203 -0.108  0.322 -0.108 -0.300   
3   ...    -0.5 -0.108 -0.076 -0.054  0.831 -0.108  0.322 -0.108 -0.300   
4   ...    -0.5 -0.108 -0.076 -0.054  0.831 -0.108 -3.101 -0.108  3.332   

   Target  
0       0  
1       0  
2       0  
3       0  
4       0  

[5 rows x 43 columns]

# Task 2 -K-NN Algorithm Solution from Fundamentals of Machine Learning Chapter 5 - Question 3
# Using Manhattan Distance

In [27]:
dataq2 = pd.read_csv('Asignment1_Q2.csv',index_col=False) # Importing Chapter 5 Q3 dataset
dataq2

COUNTRY_ID  LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS  \
0   Afghanistan     59.61         23.21         74.3       4.44           0.4   
1         Haiti     45.00         47.67         73.1       0.09           3.4   
2       Nigeria     51.30         38.23         82.6       1.07           4.1   
3         Egypt     70.48         26.58         19.6       1.86           5.3   
4     Argentina     75.77         32.30         13.3       0.76          10.1   
5         China     74.87         29.98         13.7       1.95           6.4   
6        Brazil     73.12         42.93         14.5       1.43           7.2   
7        Israel     81.30         28.80          3.6       6.77          12.5   
8         U.S.A     78.51         29.85          6.3       4.72          13.7   
9       Ireland     80.15         27.23          3.5       0.60          11.5   
10         U.K.     80.09         28.49          4.4       2.59          13.0   
11      Germany     80.24         22.07          3.5       1.31          12.0   
12       Canada     80.99         24.79          4.9       1.42          14.2   
13    Australia     82.09         25.40          4.2       1.86          11.5   
14       Sweden     81.43         22.18          2.4       1.27          12.8   
15  New Zealand     80.67         27.81          4.9       1.13          12.3   
16       Russia     67.62         31.68         10.0       3.87          12.9   

       CPI  
0   1.5171  
1   1.7999  
2   2.4493  
3   2.8622  
4   2.9961  
5   3.6356  
6   3.7741  
7   5.8069  
8   7.1357  
9    7.536  
10  7.7751  
11  8.0461  
12  8.6725  
13  8.8442  
14  9.2985  
15  9.4627  
16       ?

In [28]:
td =dataq2.iloc[0:16,:] # Subsetting descriptive dataset
td.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16 entries, 0 to 15
Data columns (total 7 columns):
COUNTRY_ID      16 non-null object
LIFE_EXP        16 non-null float64
TOP10_INCOME    16 non-null float64
INFANT_MORT     16 non-null float64
MIL_SPEND       16 non-null float64
SCHOOL_YEARS    16 non-null float64
CPI             16 non-null object
dtypes: float64(5), object(2)
memory usage: 976.0+ bytes


In [29]:
qd = np.array([67.62, 31.68, 10.00, 3.87, 12.90]) # Creating query dataset to predict target feature 'CPI'

In [30]:
features = ['LIFE_EXP','TOP10_INCOME','INFANT_MORT','MIL_SPEND','SCHOOL_YEARS']
Manhattan_distance = np.sum(abs(np.array(td[features])-qd), axis=1) # Calculating Manhattan Distance
td['Manhattan_distance']= Manhattan_distance # Creating new column
td['Manhattan_distance'].head()

0     93.85
1    114.99
2    107.07
3     27.17
4     17.98
Name: Manhattan_distance, dtype: float64

In [31]:
td = td.sort_values(by=['Manhattan_distance']) # Sorting dataset by Manhattan Distance
td['Ordered_Manhattan']=range(td.shape[0]) # Creating ranking by smallest distance
td.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16 entries, 4 to 1
Data columns (total 9 columns):
COUNTRY_ID            16 non-null object
LIFE_EXP              16 non-null float64
TOP10_INCOME          16 non-null float64
INFANT_MORT           16 non-null float64
MIL_SPEND             16 non-null float64
SCHOOL_YEARS          16 non-null float64
CPI                   16 non-null object
Manhattan_distance    16 non-null float64
Ordered_Manhattan     16 non-null int64
dtypes: float64(6), int64(1), object(2)
memory usage: 1.2+ KB


In [32]:
td['CPI'] = pd.to_numeric(td['CPI']); # Converting target feature CPI from Pandas object to float
td['CPI'].dtype

dtype('float64')

### a.) Predicted value of CPI for Russia using Manhattan Distance based 3-nearest neighbour prediction model 

In [33]:
np.round(np.mean(td.loc[td.Ordered_Manhattan<3,'CPI']),4)

4.5891

### b.) Predicted value of CPI for Russia using weighing reciprocal scheme in K=16 nearest neighbour prediction model

In [34]:
reqweight = 1 / (td['Manhattan_distance'])**2; # Calculating reqweight for each observation
np.round(sum(reqweight*td['CPI'])/sum(reqweight),4)

6.1215

### c.) Normalizing descriptive features and calculating target CPI

### Normalization

In [35]:
norm_td = td
norm_qd = qd
min_array = []
max_array = []
for i in features:
    min_val=min(norm_td[i])
    
    max_val=max(norm_td[i])
    
    norm_td[i]=(norm_td[i]-min_val)/(max_val-min_val)
    
    min_array.append(min_val)
    
    max_array.append(max_val)
    
norm_td.head()

COUNTRY_ID  LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS  \
4     Argentina  0.829604      0.399609     0.135910   0.100299      0.702899   
8         U.S.A  0.903478      0.303906     0.048628   0.693114      0.963768   
5         China  0.805338      0.308984     0.140898   0.278443      0.434783   
10         U.K.  0.946077      0.250781     0.024938   0.374251      0.913043   
15  New Zealand  0.961715      0.224219     0.031172   0.155689      0.862319   

       CPI  Manhattan_distance  Ordered_Manhattan  
4   2.9961               17.98                  0  
8   7.1357               18.07                  1  
5   3.6356               21.07                  2  
10  7.7751               22.64                  3  
15  9.4627               25.36                  4

### Normalizing data

In [36]:
norm_qd = (np.array(qd)-np.array(min_array))/(np.array(max_array)-np.array(min_array))
norm_qd

array([0.60986789, 0.37539062, 0.09476309, 0.56586826, 0.9057971 ])

### Manhattan Distance for Normalized Data

In [37]:
norm_distance = np.sum(abs(np.array(norm_td[features])-norm_qd), axis =1) # Calculation manhattan distance
norm_td['Manhattan_distance_N'] = norm_distance

###  Ordering Normalized Imput based on Manhattan Distance

In [38]:
norm_td=norm_td.sort_values(by=['Manhattan_distance_N'])
norm_td['Ordered_Manhattan_N']=range(norm_td.shape[0])
norm_td.head()

COUNTRY_ID  LIFE_EXP  TOP10_INCOME  INFANT_MORT  MIL_SPEND  SCHOOL_YEARS  \
8         U.S.A  0.903478      0.303906     0.048628   0.693114      0.963768   
10         U.K.  0.946077      0.250781     0.024938   0.374251      0.913043   
4     Argentina  0.829604      0.399609     0.135910   0.100299      0.702899   
15  New Zealand  0.961715      0.224219     0.031172   0.155689      0.862319   
7        Israel  0.978700      0.262891     0.014963   1.000000      0.876812   

       CPI  Manhattan_distance  Ordered_Manhattan  Manhattan_distance_N  \
8   7.1357               18.07                  1              0.596446   
10  7.7751               22.64                  3              0.729507   
4   2.9961               17.98                  0              0.953569   
15  9.4627               25.36                  4              1.020268   
7   5.8069               26.26                  5              1.024250   

    Ordered_Manhattan_N  
8                     0  
10                    1  
4                     2  
15                    3  
7                     4

### Question 3

In [39]:
np.round(np.mean(norm_td.loc[norm_td.Ordered_Manhattan_N<3,'CPI']),4)

5.969

### Question 4

In [40]:
reqweight_norm = 1 / (norm_td['Manhattan_distance_N'])**2
np.round(sum(reqweight_norm*norm_td['CPI'])/sum(reqweight_norm),4)

6.6091

### Question 5 - Best Model selection

#### Ans: We would select K=3 NN algorithm based on weighting scheme whose predicted value is 5.969.  There are two main reasons. 
#### 1.)  Normalization of descritive features and target feature is necessary to bring all features in same scale as the data rages are varied widely among each other. 
#### 2.) K=3 is chosen as the data is densely population and gives better result and weighted model gives a similar result for this model. Thus, considering computation time and model complexity, it is preferable to chose 3-NN normalized model.
#### However, we notice that k=3 model without normalization is giving a better result and is closest but just one prediction cannot prove the accuracy and thus more accuraicy measures should be used .  